Sources/Basic/SyncronizedQueue.swift

```diff
@@ -0,0 +1,58 @@
+/*
+ This source file is part of the Swift.org open source project
+
+ Copyright 2015 - 2016 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See http://swift.org/LICENSE.txt for license information
+ See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+*/
+
+import class Foundation.Condition
+
+/// This class can be used as a shared queue between multiple threads providing
+/// thread safe APIs. The elements are stored as optionals so nil can be used as
+/// a sentinel.
+public final class SyncronizedQueue<Element> {
+
+    /// Storage for queued elements.
+    private var storage: [Element?]
+
+    /// Condition variable to block the thread trying dequeue and queue is empty.
+    private var notEmptyCondition: Condition
+
+    /// Create a default instance of queue.
+    public init() {
+        storage = []
+        notEmptyCondition = Condition()
+    }
+
+    /// Safely enqueue an element to end of the queue and signals a thread blocked on dequeue.
+    ///
+    /// - Parameters:
+    ///     - element: The element to be enqueued.
+    public func enqueue(_ element: Element?) {
+        notEmptyCondition.lock()
```

Comment:
Should we get an extension on Condition to add something like `whileLocked` taking a closure, to replace this lock + defer pattern?